src/Dumplie/Domain/Customer/Cart.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+declare (strict_types = 1);
+
+namespace Dumplie\Domain\Customer;
+
+use Dumplie\Domain\Customer\Exception\ProductNotAvailableException;
+use Dumplie\Domain\SharedKernel\Exception\InvalidCurrencyException;
+use Dumplie\Domain\SharedKernel\Money\Price;
+
+final class Cart
+{
+    /**
+     * @var CartId
+     */
+    private $id;
+
+    /**
+     * @var CartItem[]|array
+     */
+    private $items;
+
+    public function __construct()
+    {
+        $this->id = CartId::generate();
+        $this->items = [];
+    }
+
+    /**
+     * @return CartId
+     */
+    public function id() : CartId
+    {
+        return $this->id;
+    }
+
+    /**
+     * @param Product $product
+     * @param int $quantity
+     * @throws \Dumplie\Domain\SharedKernel\Exception\InvalidCurrencyException
+     * @throws \Dumplie\Domain\Customer\Exception\ProductNotAvailableException
+     */
+    public function add(Product $product, int $quantity)
+    {
+        if (!$product->isAvailable()) {
+            throw new ProductNotAvailableException($product);
+        }
+
+        foreach ($this->items as $item) {
```

Comment:
why checking with all items, when u could [only with the first one](https://github.com/dumplie/dumplie/pull/1/files#diff-d328ba413ca672bc7f41499312843260R61)?

Plus, for micro–optimization, it's necessary only the first time a `Product` is added to the `Cart`.
So, if u still want to check all items, do it in constructor.